### ISSUE TITLE
option to trim leading spaces

### ISSUE THREAD

**Issue by joprice** [bloomberg/ppx_string_interpolation#9]
**Is your feature request related to a problem? Please describe.**

When leading spaces are significant, the text must be fully justified to the left of the file to avoid including them:

```ocaml
   let ssdp_message =
      [%string
        {|M-SEARCH * HTTP/1.1
HOST:$group
MAN:ssdp:discover
MX:10
ST:urn:smartspeaker-audio:service:SpeakerGroup:1
|}]
``` 

**Describe the solution you'd like**

A separate extension or a special character (as the pipe character in multiline strings in scala) could be used to indicate the leading space should be removed

```ocaml
   let ssdp_message =
      [%string.trimmed
        {|M-SEARCH * HTTP/1.1
          HOST:$group
          MAN:ssdp:discover
          MX:10
          ST:urn:smartspeaker-audio:service:SpeakerGroup:1
        |}]
``` 

or

```ocaml
   let ssdp_message =
      [%string
        {|M-SEARCH * HTTP/1.1
          |HOST:$group
          |MAN:ssdp:discover
          |MX:10
          |ST:urn:smartspeaker-audio:service:SpeakerGroup:1
        |}]
```

**Comment by ksromanov** [comment]
Thanks for the suggestion. Yes, it makes sense.This is especially useful for C++ codegeneration like

```ocaml
let emptyTemplateTClass name =
  [%string
     {|
      template <class T> class $name {
          $name() {}
          ~$name() {}
      };
      |}]
```

I am not sure though if .trimmed is a good suffix.

**Comment by joprice** [comment]
I agree. I didn't like it when I wrote it. `string.indented` sounds good, also `string.strip_leading`, `string.remove_padding`, `string.strip_margin`. The last one is the terminology scala uses, but more specifically means that there is a character that is used to designate where the margin starts, which means the whitespace can differ per line.  

Also, I haven't checked the implementation of the ppx to know if this could happen at compile time, but the scala version I mentioned is actually a runtime implementation: https://github.com/scala/scala/blob/v2.13.6/src/library/scala/collection/StringOps.scala#L765-L784. If these strings are created in a tight loop, maybe you would benefit from the compiler preprocessing them, but if they are created once at startup, then this can just be a library function instead and doesn't have to be in a ppx.

**Comment by ksromanov** [comment]
I would prefer to use the name without verb, since it makes it lengthy and harder to type (people usually want to type less, if possible). So, `string.padded` or smth like that?

If it is possible, i.e. there are no blockers, which I do not see right now, it should be compile-time implementation, since it is even easier to implement than run-time.

I am mostly worried about the interaction with the ecosystem, especially formatters, which can rewrite many things and move the leading `{|`. I would prefer to have a clean simple semantics + minimum of visual noise. Therefore your first variant is better than `Scala`-like, though the latter is "formatter-proof".

I will think about this sometime next week if you do not mind - we are quite busy now.

**Comment by joprice** [comment]
No urgency on my end. It would be a convenient addition but not strictly necessary.

**Comment by SoulPancake** [comment]
@ksromanov Do we need to still pick this up?

**Comment by ksromanov** [comment]
I think not.
